Support recursive definitions in JS

Currently, recursive definitions don't work:

```ts
:js
rec def foo =
  let _ = console.log "hey" in foo
//│ // Query 1
//│ let foo1;
//│ foo1 = (function (_) {
//│   return foo;
//│ })(console.log("hey"));
//│ res = foo1;
//│ // End of generated code
//│ foo: nothing
//│    = [Function: foo]

:js
rec def foo() =
  let _ = console.log "hey" in foo()
//│ // Query 1
//│ let foo2;
//│ foo2 = () => (function (_) {
//│   return foo1();
//│ })(console.log("hey"));
//│ res = foo2;
//│ // End of generated code
//│ foo: () -> nothing
//│    = [Function: foo2]
```

Weirdly, sometimes they do seem to work:

```ts

// TODO define concat
:e
def concat: string -> string -> string
concat s t = s + t
//│ concat: string -> string -> string
//│ int -> int -> int
//│   <:  concat:
//│ string -> string -> string
//│ ╔══[ERROR] Type mismatch in def definition:
//│ ║  l.56: 	concat s t = s + t
//│ ║        	^^^^^^^^^^^^^^^^^^
//│ ╟── expression of type `string` does not match type `int`
//│ ║  l.55: 	def concat: string -> string -> string
//│ ║        	            ^^^^^^
//│ ╟── Note: constraint arises from reference:
//│ ║  l.56: 	concat s t = s + t
//│ ╙──      	             ^
//│ ╔══[ERROR] Type mismatch in def definition:
//│ ║  l.56: 	concat s t = s + t
//│ ║        	^^^^^^^^^^^^^^^^^^
//│ ╟── expression of type `int` does not match type `string`
//│ ║  l.56: 	concat s t = s + t
//│ ║        	             ^^^^^
//│ ╟── but it flows into function of type `?a -> ?b`
//│ ║  l.56: 	concat s t = s + t
//│ ║        	         ^^^^^^^^^
//│ ╟── which does not match type `string -> string`
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.55: 	def concat: string -> string -> string
//│ ╙──      	                                ^^^^^^
//│       = [Function: concat]

:js
rec def test n acc =
  if n == 0 then acc
  else concat (test (n - 1) (concat acc "+")) "!"
//│ // Query 1
//│ let test;
//│ test = (n) => (acc) => (n == 0 ? acc : concat(test(n - 1)(concat(acc)("+")))("!"));
//│ res = test;
//│ // End of generated code
//│ test: int -> string -> string
//│     = [Function: test]

test 1 "hello"
//│ res: string
//│    = 'hello+!'

test 10 "hello"
//│ res: string
//│    = 'hello++++++++++!!!!!!!!!!'
```
Is this inconsistency a bug in the handling of scopes/names?

Other examples:
```ts
:js
def pow x =
  let rec go n =
    if n > 0 then go (n - 1) * x
    else 1
  in go
//│ // Query 1
//│ let pow;
//│ pow = (x) => (function (go) {
//│   return go;
//│ })((n) => (n > 0 ? go(n - 1) * x : 1));
//│ res = pow;
//│ // End of generated code
//│ pow: int -> int -> int
//│    = [Function: pow]

pow 3 4
//│ res: int
//│ Unexpected runtime error
//│   ReferenceError: go is not defined

:js
rec def pow n = { call = fun x -> if n > 0 then (pow (n - 1)).call x * x else 1 }
//│ // Query 1
//│ let pow1;
//│ pow1 = (n) => { call: (x) => (n > 0 ? pow(n - 1).call(x) * x : 1) };
//│ res = pow1;
//│ // End of generated code
//│ pow: int -> {call: int -> int}
//│    = [Function: pow1]

(pow 3).call 4
//│ res: int
//│ Unexpected runtime error
//│   TypeError: Cannot read properties of undefined (reading 'call')
```

--- a/shared/src/main/scala/mlscript/JSBackend.scala
+++ b/shared/src/main/scala/mlscript/JSBackend.scala
@@ -129,16 +129,33 @@ class JSBackend {
     case Sel(receiver, fieldName) =>
       JSField(translateTerm(receiver), fieldName.name)
     // Turn let into an IIFE.
-    case Let(isRec, Var(name), value, body) =>
+    case Let(true, Var(name), Lam(args, body), expr) =>
+      val letScope = scope.derive("Let", name :: Nil)
+      val fn = {
+        val params = translateParams(args)
+        val bindings = name :: params.flatMap { _.bindings }
+        val fnScope = scope.derive("Function", bindings)
+        val fnBody = fnScope.tempVars.`with`(translateTerm(body)(fnScope))
+        JSFuncExpr(S(name), params, fnBody.fold(_.`return` :: Nil, identity))
+      }
+      JSImmEvalFn(
+        N,
+        JSNamePattern(name) :: Nil,
+        letScope.tempVars.`with`(translateTerm(expr)(letScope)),
+        fn :: Nil
+      )
+    case Let(_, Var(name), value, body) =>
       val letScope = scope.derive("Let", name :: Nil)
       JSImmEvalFn(
-        name :: Nil,
+        N,
+        JSNamePattern(name) :: Nil,
         letScope.tempVars `with` translateTerm(body)(letScope),
         translateTerm(value)(letScope) :: Nil
       )
     case Blk(stmts) =>
       val blkScope = scope.derive("Blk")
       JSImmEvalFn(
+        N,
         Nil,
         R(blkScope.tempVars `with` (stmts flatMap (_.desugared._2) map {
           case t: Term             => JSExprStmt(translateTerm(t))
@@ -195,7 +212,7 @@ class JSBackend {
     case Case(pat, body, rest) =>
       translateCase(scrut, pat)(translateTerm(body), translateCaseBranch(scrut, rest))
     case Wildcard(body) => translateTerm(body)
-    case NoCases        => JSImmEvalFn(Nil, R(JSInvoke(
+    case NoCases        => JSImmEvalFn(N, Nil, R(JSInvoke(
       JSNew(JSIdent("Error")),
       JSExpr("non-exhaustive case expression") :: Nil
     ).`throw` :: Nil), Nil)
@@ -360,7 +377,7 @@ class JSWebBackend extends JSBackend {
             ).stmt :: Nil
         })
     val epilogue = resultsIdent.member("map")(JSIdent(prettyPrinterName)).`return` :: Nil
-    JSImmEvalFn(Nil, R(polyfill.emit() ::: stmts ::: epilogue), Nil).toSourceCode.toLines
+    JSImmEvalFn(N, Nil, R(polyfill.emit() ::: stmts ::: epilogue), Nil).toSourceCode.toLines
   }
 }
 

--- a/shared/src/main/scala/mlscript/JSBackend.scala
+++ b/shared/src/main/scala/mlscript/JSBackend.scala
@@ -145,7 +145,7 @@ class JSBackend {
         fn :: Nil
       )
     case Let(true, Var(name), _, _) =>
-      throw new CodeGenError(s"recursive value $name is not supported")
+      throw new CodeGenError(s"recursive non-function definition $name is not supported")
     case Let(_, Var(name), value, body) =>
       val letScope = scope.derive("Let", name :: Nil)
       JSImmEvalFn(

--- a/shared/src/main/scala/mlscript/JSBackend.scala
+++ b/shared/src/main/scala/mlscript/JSBackend.scala
@@ -144,13 +144,15 @@ class JSBackend {
         letScope.tempVars.`with`(translateTerm(expr)(letScope)),
         fn :: Nil
       )
+    case Let(true, Var(name), _, _) =>
+      throw new CodeGenError(s"recursive value $name is not supported")
     case Let(_, Var(name), value, body) =>
       val letScope = scope.derive("Let", name :: Nil)
       JSImmEvalFn(
         N,
         JSNamePattern(name) :: Nil,
         letScope.tempVars `with` translateTerm(body)(letScope),
-        translateTerm(value)(letScope) :: Nil
+        translateTerm(value) :: Nil
       )
     case Blk(stmts) =>
       val blkScope = scope.derive("Blk")

--- a/shared/src/main/scala/mlscript/codegen/Codegen.scala
+++ b/shared/src/main/scala/mlscript/codegen/Codegen.scala
@@ -418,13 +418,15 @@ final case class JSFuncExpr(name: Opt[Str], params: Ls[JSPattern], body: Ls[JSSt
 
 // IIFE: immediately invoked function expression
 final case class JSImmEvalFn(
-    params: Ls[Str],
+    name: Opt[Str],
+    params: Ls[JSPattern],
     body: Either[JSExpr, Ls[JSStmt]],
     arguments: Ls[JSExpr]
 ) extends JSExpr {
   implicit def precedence: Int = 22
   def toSourceCode: SourceCode = {
-    (SourceCode(s"function (${params mkString ", "}) ") ++ (body match {
+    val fnName = name.getOrElse("")
+    (SourceCode(s"function $fnName${JSExpr.params(params)} ") ++ (body match {
       case Left(expr) => new JSReturnStmt(expr).toSourceCode
       case Right(stmts) =>
         stmts.foldLeft(SourceCode.empty) { _ + _.toSourceCode }

--- a/shared/src/main/scala/mlscript/codegen/Helpers.scala
+++ b/shared/src/main/scala/mlscript/codegen/Helpers.scala
@@ -23,7 +23,7 @@ object Helpers {
         .mkString(", ")
       s"Rcd($entries})"
     case Sel(receiver, fieldName)    => s"Sel(${inspect(receiver)}, $fieldName)"
-    case Let(isRec, name, rhs, body) => s"Let($isRec, $name)"
+    case Let(isRec, name, rhs, body) => s"Let($isRec, $name, ${inspect(rhs)}, ${inspect(body)})"
     case Blk(stmts)                  => s"Blk(...)"
     case Bra(rcd, trm)               => s"Bra(rcd = $rcd, ${inspect(trm)})"
     case Asc(trm, ty)                => s"Asc(${inspect(trm)}, $ty)"

--- a/shared/src/test/diff/codegen/Recursion.mls
+++ b/shared/src/test/diff/codegen/Recursion.mls
@@ -54,3 +54,15 @@ rec def pow2 n = { call = fun x -> if n > 0 then (pow2 (n - 1)).call x * x else 
 (pow2 3).call 4
 //│ res: int
 //│    = 64
+
+:js
+let rec x = { f = x } in x
+//│ res: {f: 'a} as 'a
+//│ Code generation encountered an error:
+//│   recursive value x is not supported
+
+:js
+let x = { f = x } in x
+//│ res: {f: 'a} as 'a
+//│ Code generation encountered an error:
+//│   unresolved symbol x

--- a/shared/src/test/diff/codegen/Recursion.mls
+++ b/shared/src/test/diff/codegen/Recursion.mls
@@ -39,7 +39,6 @@ def pow1 x =
 //│ pow1: int -> int -> int
 //│     = [Function: pow1]
 
-// FIXME support recursive let bindings
 pow1 3 4
 //│ res: int
 //│    = 81

--- a/shared/src/test/diff/codegen/Recursion.mls
+++ b/shared/src/test/diff/codegen/Recursion.mls
@@ -18,16 +18,21 @@ pow0 3 4
 //│    = 64
 
 :js
+:p
 def pow1 x =
   let rec go n =
     if n > 0 then go (n - 1) * x
     else 1
   in go
+//│ Parsed: def pow1: ((x,) => (let rec go = ((n,) => (((if ((> (n,)) (0,))) ((* ((go ((((- (n,)) (1,))),)),)) (x,))) 1)); go));
+//│ Desugared: def pow1: ((x,) => (let rec go = ((n,) => (((if ((> (n,)) (0,))) ((* ((go ((((- (n,)) (1,))),)),)) (x,))) 1)); go))
 //│ // Query 1
 //│ globalThis.pow1 = function pow1(x) {
 //│   return ((function (go) {
 //│     return go;
-//│   })((n) => n > 0 ? go(n - 1) * x : 1));
+//│   })(function go(n) {
+//│     return n > 0 ? go(n - 1) * x : 1;
+//│   }));
 //│ };
 //│ res = pow1;
 //│ // End of generated code
@@ -37,8 +42,7 @@ def pow1 x =
 // FIXME support recursive let bindings
 pow1 3 4
 //│ res: int
-//│ Runtime error:
-//│   ReferenceError: go is not defined
+//│    = 81
 
 :js
 rec def pow2 n = { call = fun x -> if n > 0 then (pow2 (n - 1)).call x * x else 1 }

--- a/shared/src/test/diff/codegen/Recursion.mls
+++ b/shared/src/test/diff/codegen/Recursion.mls
@@ -59,7 +59,7 @@ rec def pow2 n = { call = fun x -> if n > 0 then (pow2 (n - 1)).call x * x else 
 let rec x = { f = x } in x
 //│ res: {f: 'a} as 'a
 //│ Code generation encountered an error:
-//│   recursive value x is not supported
+//│   recursive non-function definition x is not supported
 
 :js
 let x = { f = x } in x

--- a/shared/src/test/diff/codegen/Recursion.mls
+++ b/shared/src/test/diff/codegen/Recursion.mls
@@ -18,14 +18,11 @@ pow0 3 4
 //│    = 64
 
 :js
-:p
 def pow1 x =
   let rec go n =
     if n > 0 then go (n - 1) * x
     else 1
   in go
-//│ Parsed: def pow1: ((x,) => (let rec go = ((n,) => (((if ((> (n,)) (0,))) ((* ((go ((((- (n,)) (1,))),)),)) (x,))) 1)); go));
-//│ Desugared: def pow1: ((x,) => (let rec go = ((n,) => (((if ((> (n,)) (0,))) ((* ((go ((((- (n,)) (1,))),)),)) (x,))) 1)); go))
 //│ // Query 1
 //│ globalThis.pow1 = function pow1(x) {
 //│   return ((function (go) {

--- a/shared/src/test/diff/mlscript/RecursiveTypes.mls
+++ b/shared/src/test/diff/mlscript/RecursiveTypes.mls
@@ -284,16 +284,16 @@ bar2_ty2 = bar_ty2
 //│   <:  bar2_ty2:
 //│ {x: 'a} as 'a
 //│ ╔══[ERROR] Type mismatch in def definition:
-//│ ║  l.269: 	bar2_ty2 = bar_ty2
+//│ ║  l.282: 	bar2_ty2 = bar_ty2
 //│ ║         	^^^^^^^^^^^^^^^^^^
 //│ ╟── type `C[?r]` does not match type `'r | {x: ?r0}`
-//│ ║  l.205: 	def bar_ty2: C['r] as 'r
+//│ ║  l.218: 	def bar_ty2: C['r] as 'r
 //│ ║         	             ^^^^^
 //│ ╟── but it flows into reference with expected type `'r | {x: ?r1}`
-//│ ║  l.269: 	bar2_ty2 = bar_ty2
+//│ ║  l.282: 	bar2_ty2 = bar_ty2
 //│ ║         	           ^^^^^^^
 //│ ╟── Note: constraint arises from local type binding:
-//│ ║  l.250: 	def bar2_ty2: { x: 'r } as 'r
+//│ ║  l.263: 	def bar2_ty2: { x: 'r } as 'r
 //│ ╙──       	              ^^^^^^^^^
 //│ Runtime error:
 //│   ReferenceError: bar_ty21 is not defined
@@ -304,40 +304,40 @@ bar_ty2 = bar2_ty2
 //│   <:  bar_ty2:
 //│ 'r | (C[(C['c .. 'd] with {a: 'b}) & 'r as 'b .. 'a] with {a: 'a}) as 'a
 //│ ╔══[ERROR] Type mismatch in def definition:
-//│ ║  l.289: 	bar_ty2 = bar2_ty2
+//│ ║  l.302: 	bar_ty2 = bar2_ty2
 //│ ║         	^^^^^^^^^^^^^^^^^^
 //│ ╟── type `{x: ?r}` does not match type `'r | C[?r0]`
-//│ ║  l.250: 	def bar2_ty2: { x: 'r } as 'r
+//│ ║  l.263: 	def bar2_ty2: { x: 'r } as 'r
 //│ ║         	              ^^^^^^^^^
 //│ ╟── but it flows into reference with expected type `'r | C[?r1]`
-//│ ║  l.289: 	bar_ty2 = bar2_ty2
+//│ ║  l.302: 	bar_ty2 = bar2_ty2
 //│ ║         	          ^^^^^^^^
 //│ ╟── Note: constraint arises from local type binding:
-//│ ║  l.205: 	def bar_ty2: C['r] as 'r
+//│ ║  l.218: 	def bar_ty2: C['r] as 'r
 //│ ╙──       	             ^^^^^
 //│ ╔══[ERROR] Type mismatch in def definition:
-//│ ║  l.289: 	bar_ty2 = bar2_ty2
+//│ ║  l.302: 	bar_ty2 = bar2_ty2
 //│ ║         	^^^^^^^^^^^^^^^^^^
 //│ ╟── type `{x: ?r}` does not match type `'r | C[?r0]`
-//│ ║  l.250: 	def bar2_ty2: { x: 'r } as 'r
+//│ ║  l.263: 	def bar2_ty2: { x: 'r } as 'r
 //│ ║         	              ^^^^^^^^^
 //│ ╟── but it flows into reference with expected type `'r | C[?r1]`
-//│ ║  l.289: 	bar_ty2 = bar2_ty2
+//│ ║  l.302: 	bar_ty2 = bar2_ty2
 //│ ║         	          ^^^^^^^^
 //│ ╟── Note: constraint arises from local type binding:
-//│ ║  l.205: 	def bar_ty2: C['r] as 'r
+//│ ║  l.218: 	def bar_ty2: C['r] as 'r
 //│ ╙──       	             ^^^^^
 //│ ╔══[ERROR] Type mismatch in def definition:
-//│ ║  l.289: 	bar_ty2 = bar2_ty2
+//│ ║  l.302: 	bar_ty2 = bar2_ty2
 //│ ║         	^^^^^^^^^^^^^^^^^^
 //│ ╟── type `{x: ?r}` does not match type `'r | C[?r0]`
-//│ ║  l.250: 	def bar2_ty2: { x: 'r } as 'r
+//│ ║  l.263: 	def bar2_ty2: { x: 'r } as 'r
 //│ ║         	              ^^^^^^^^^
 //│ ╟── but it flows into reference with expected type `'r | C[?r1]`
-//│ ║  l.289: 	bar_ty2 = bar2_ty2
+//│ ║  l.302: 	bar_ty2 = bar2_ty2
 //│ ║         	          ^^^^^^^^
 //│ ╟── Note: constraint arises from local type binding:
-//│ ║  l.205: 	def bar_ty2: C['r] as 'r
+//│ ║  l.218: 	def bar_ty2: C['r] as 'r
 //│ ╙──       	             ^^^^^
 //│ Runtime error:
 //│   ReferenceError: bar2_ty21 is not defined
@@ -375,13 +375,13 @@ f
 :e
 f 1
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.363: 	f 1
+//│ ║  l.376: 	f 1
 //│ ║         	^^^
 //│ ╟── operator application of type `int` does not have field 'a'
-//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.361: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ║         	                                             ^^^^^
 //│ ╟── Note: constraint arises from field selection:
-//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.361: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ╙──       	                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 //│ res: error | int | (('b | 1 | 'a)\a & {a: int} as 'a)
 //│ Runtime error:
@@ -390,13 +390,13 @@ f 1
 :e
 f { a = 1 }
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.378: 	f { a = 1 }
+//│ ║  l.391: 	f { a = 1 }
 //│ ║         	^^^^^^^^^^^
 //│ ╟── operator application of type `int` does not have field 'a'
-//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.361: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ║         	                                             ^^^^^
 //│ ╟── Note: constraint arises from field selection:
-//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.361: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ╙──       	                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 //│ res: error
 //│    = { a: 1 }
@@ -414,13 +414,13 @@ f ainf
 //│ Runtime error:
 //│   ReferenceError: ainf is not defined
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.399: 	f ainf
+//│ ║  l.412: 	f ainf
 //│ ║         	^^^^^^
 //│ ╟── operator application of type `int` does not have field 'a'
-//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.361: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ║         	                                             ^^^^^
 //│ ╟── Note: constraint arises from field selection:
-//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.361: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ╙──       	                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 //│ res: error
 //│ Runtime error:
@@ -433,13 +433,13 @@ f infina
 //│ Runtime error:
 //│   ReferenceError: infina is not defined
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.418: 	f infina
+//│ ║  l.431: 	f infina
 //│ ║         	^^^^^^^^
 //│ ╟── operator application of type `int` does not have field 'a'
-//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.361: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ║         	                                             ^^^^^
 //│ ╟── Note: constraint arises from field selection:
-//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.361: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ╙──       	                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 //│ res: error | int | (('b | (0 & {a: 'c} as 'c) | 'a)\a & {a: int} as 'a)
 //│ Runtime error:
@@ -453,16 +453,16 @@ def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as
 :e
 f_manual 1
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.441: 	f_manual 1
+//│ ║  l.454: 	f_manual 1
 //│ ║         	^^^^^^^^^^
 //│ ╟── integer literal of type `1` does not have field 'a'
-//│ ║  l.441: 	f_manual 1
+//│ ║  l.454: 	f_manual 1
 //│ ║         	         ^
 //│ ╟── Note: constraint arises from record type:
-//│ ║  l.436: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
+//│ ║  l.449: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
 //│ ║         	                ^^^^^^^^^^^^^^^^^
 //│ ╟── from intersection type:
-//│ ║  l.436: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
+//│ ║  l.449: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
 //│ ╙──       	              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 //│ res: 1 | error | (('b | 1 | 'a)\a & {a: int} as 'a)
 //│    = <no result>
@@ -471,16 +471,16 @@ f_manual 1
 :e
 f_manual { a = 1 }
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.459: 	f_manual { a = 1 }
+//│ ║  l.472: 	f_manual { a = 1 }
 //│ ║         	^^^^^^^^^^^^^^^^^^
 //│ ╟── integer literal of type `1` does not have field 'a'
-//│ ║  l.459: 	f_manual { a = 1 }
+//│ ║  l.472: 	f_manual { a = 1 }
 //│ ║         	               ^
 //│ ╟── Note: constraint arises from record type:
-//│ ║  l.436: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
+//│ ║  l.449: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
 //│ ║         	                ^^^^^^^^^^^^^^^^^
 //│ ╟── from intersection type:
-//│ ║  l.436: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
+//│ ║  l.449: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
 //│ ╙──       	                    ^^^^^^^^^^^^
 //│ res: 1 | error | ({a: 1} | 'a\a & {a: int} as 'a)
 //│    = <no result>
@@ -507,16 +507,16 @@ def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & ((
 :e
 f_manual_ns ainf
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.495: 	f_manual_ns ainf
+//│ ║  l.508: 	f_manual_ns ainf
 //│ ║         	^^^^^^^^^^^^^^^^
 //│ ╟── type `int` does not have field 'a'
-//│ ║  l.490: 	def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & (({a: 'd & 'c} as 'c) | ~{a: 'e | int})\a & (({a: 'f} as 'c) as 'f) & (int | ~{a: 'e | int} | ~{})\a & (int | ~{a: 'e | int})\a & int & int) -> ('g | 'd | ('b | 'h\a & {a: 'e | int} as 'h))
+//│ ║  l.503: 	def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & (({a: 'd & 'c} as 'c) | ~{a: 'e | int})\a & (({a: 'f} as 'c) as 'f) & (int | ~{a: 'e | int} | ~{})\a & (int | ~{a: 'e | int})\a & int & int) -> ('g | 'd | ('b | 'h\a & {a: 'e | int} as 'h))
 //│ ║         	                                                              ^^^
 //│ ╟── Note: constraint arises from record type:
-//│ ║  l.490: 	def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & (({a: 'd & 'c} as 'c) | ~{a: 'e | int})\a & (({a: 'f} as 'c) as 'f) & (int | ~{a: 'e | int} | ~{})\a & (int | ~{a: 'e | int})\a & int & int) -> ('g | 'd | ('b | 'h\a & {a: 'e | int} as 'h))
+//│ ║  l.503: 	def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & (({a: 'd & 'c} as 'c) | ~{a: 'e | int})\a & (({a: 'f} as 'c) as 'f) & (int | ~{a: 'e | int} | ~{})\a & (int | ~{a: 'e | int})\a & int & int) -> ('g | 'd | ('b | 'h\a & {a: 'e | int} as 'h))
 //│ ║         	                              ^^^^^^^^^^^^
 //│ ╟── from intersection type:
-//│ ║  l.490: 	def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & (({a: 'd & 'c} as 'c) | ~{a: 'e | int})\a & (({a: 'f} as 'c) as 'f) & (int | ~{a: 'e | int} | ~{})\a & (int | ~{a: 'e | int})\a & int & int) -> ('g | 'd | ('b | 'h\a & {a: 'e | int} as 'h))
+//│ ║  l.503: 	def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & (({a: 'd & 'c} as 'c) | ~{a: 'e | int})\a & (({a: 'f} as 'c) as 'f) & (int | ~{a: 'e | int} | ~{})\a & (int | ~{a: 'e | int})\a & int & int) -> ('g | 'd | ('b | 'h\a & {a: 'e | int} as 'h))
 //│ ╙──       	                                  ^^^^^^^
 //│ res: error
 //│    = <no result>
@@ -550,13 +550,13 @@ r + 1
 :e
 r.a
 //│ ╔══[ERROR] Type mismatch in field selection:
-//│ ║  l.538: 	r.a
+//│ ║  l.551: 	r.a
 //│ ║         	^^^
 //│ ╟── integer literal of type `1` does not have field 'a'
-//│ ║  l.527: 	r = f 1
+//│ ║  l.540: 	r = f 1
 //│ ║         	      ^
 //│ ╟── but it flows into reference with expected type `{a: ?a}`
-//│ ║  l.538: 	r.a
+//│ ║  l.551: 	r.a
 //│ ╙──       	^
 //│ res: error | int
 //│ Runtime error:

--- a/shared/src/test/diff/mlscript/RecursiveTypes.mls
+++ b/shared/src/test/diff/mlscript/RecursiveTypes.mls
@@ -2,9 +2,23 @@
 :AllowRuntimeErrors
 
 // From Simple-sub tests:
+:js
 let rec l = fun a -> fun a -> fun a -> l in let rec r = fun a -> fun a -> r in if true then l else r
+//│ // Prelude
+//│ let res;
+//│ // Query 1
+//│ res = (function (l) {
+//│   return ((function (r) {
+//│     return true ? l : r;
+//│   })(function r(a) {
+//│     return (a) => r;
+//│   }));
+//│ })(function l(a) {
+//│   return (a) => (a) => l;
+//│ });
+//│ // End of generated code
 //│ res: anything -> anything -> anything -> anything -> anything -> anything -> 'a as 'a
-//│    = [Function (anonymous)]
+//│    = [Function: l]
 
 
 
@@ -270,16 +284,16 @@ bar2_ty2 = bar_ty2
 //│   <:  bar2_ty2:
 //│ {x: 'a} as 'a
 //│ ╔══[ERROR] Type mismatch in def definition:
-//│ ║  l.268: 	bar2_ty2 = bar_ty2
+//│ ║  l.269: 	bar2_ty2 = bar_ty2
 //│ ║         	^^^^^^^^^^^^^^^^^^
 //│ ╟── type `C[?r]` does not match type `'r | {x: ?r0}`
-//│ ║  l.204: 	def bar_ty2: C['r] as 'r
+//│ ║  l.205: 	def bar_ty2: C['r] as 'r
 //│ ║         	             ^^^^^
 //│ ╟── but it flows into reference with expected type `'r | {x: ?r1}`
-//│ ║  l.268: 	bar2_ty2 = bar_ty2
+//│ ║  l.269: 	bar2_ty2 = bar_ty2
 //│ ║         	           ^^^^^^^
 //│ ╟── Note: constraint arises from local type binding:
-//│ ║  l.249: 	def bar2_ty2: { x: 'r } as 'r
+//│ ║  l.250: 	def bar2_ty2: { x: 'r } as 'r
 //│ ╙──       	              ^^^^^^^^^
 //│ Runtime error:
 //│   ReferenceError: bar_ty21 is not defined
@@ -290,40 +304,40 @@ bar_ty2 = bar2_ty2
 //│   <:  bar_ty2:
 //│ 'r | (C[(C['c .. 'd] with {a: 'b}) & 'r as 'b .. 'a] with {a: 'a}) as 'a
 //│ ╔══[ERROR] Type mismatch in def definition:
-//│ ║  l.288: 	bar_ty2 = bar2_ty2
+//│ ║  l.289: 	bar_ty2 = bar2_ty2
 //│ ║         	^^^^^^^^^^^^^^^^^^
 //│ ╟── type `{x: ?r}` does not match type `'r | C[?r0]`
-//│ ║  l.249: 	def bar2_ty2: { x: 'r } as 'r
+//│ ║  l.250: 	def bar2_ty2: { x: 'r } as 'r
 //│ ║         	              ^^^^^^^^^
 //│ ╟── but it flows into reference with expected type `'r | C[?r1]`
-//│ ║  l.288: 	bar_ty2 = bar2_ty2
+//│ ║  l.289: 	bar_ty2 = bar2_ty2
 //│ ║         	          ^^^^^^^^
 //│ ╟── Note: constraint arises from local type binding:
-//│ ║  l.204: 	def bar_ty2: C['r] as 'r
+//│ ║  l.205: 	def bar_ty2: C['r] as 'r
 //│ ╙──       	             ^^^^^
 //│ ╔══[ERROR] Type mismatch in def definition:
-//│ ║  l.288: 	bar_ty2 = bar2_ty2
+//│ ║  l.289: 	bar_ty2 = bar2_ty2
 //│ ║         	^^^^^^^^^^^^^^^^^^
 //│ ╟── type `{x: ?r}` does not match type `'r | C[?r0]`
-//│ ║  l.249: 	def bar2_ty2: { x: 'r } as 'r
+//│ ║  l.250: 	def bar2_ty2: { x: 'r } as 'r
 //│ ║         	              ^^^^^^^^^
 //│ ╟── but it flows into reference with expected type `'r | C[?r1]`
-//│ ║  l.288: 	bar_ty2 = bar2_ty2
+//│ ║  l.289: 	bar_ty2 = bar2_ty2
 //│ ║         	          ^^^^^^^^
 //│ ╟── Note: constraint arises from local type binding:
-//│ ║  l.204: 	def bar_ty2: C['r] as 'r
+//│ ║  l.205: 	def bar_ty2: C['r] as 'r
 //│ ╙──       	             ^^^^^
 //│ ╔══[ERROR] Type mismatch in def definition:
-//│ ║  l.288: 	bar_ty2 = bar2_ty2
+//│ ║  l.289: 	bar_ty2 = bar2_ty2
 //│ ║         	^^^^^^^^^^^^^^^^^^
 //│ ╟── type `{x: ?r}` does not match type `'r | C[?r0]`
-//│ ║  l.249: 	def bar2_ty2: { x: 'r } as 'r
+//│ ║  l.250: 	def bar2_ty2: { x: 'r } as 'r
 //│ ║         	              ^^^^^^^^^
 //│ ╟── but it flows into reference with expected type `'r | C[?r1]`
-//│ ║  l.288: 	bar_ty2 = bar2_ty2
+//│ ║  l.289: 	bar_ty2 = bar2_ty2
 //│ ║         	          ^^^^^^^^
 //│ ╟── Note: constraint arises from local type binding:
-//│ ║  l.204: 	def bar_ty2: C['r] as 'r
+//│ ║  l.205: 	def bar_ty2: C['r] as 'r
 //│ ╙──       	             ^^^^^
 //│ Runtime error:
 //│   ReferenceError: bar2_ty21 is not defined
@@ -361,13 +375,13 @@ f
 :e
 f 1
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.362: 	f 1
+//│ ║  l.363: 	f 1
 //│ ║         	^^^
 //│ ╟── operator application of type `int` does not have field 'a'
-//│ ║  l.347: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ║         	                                             ^^^^^
 //│ ╟── Note: constraint arises from field selection:
-//│ ║  l.347: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ╙──       	                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 //│ res: error | int | (('b | 1 | 'a)\a & {a: int} as 'a)
 //│ Runtime error:
@@ -376,13 +390,13 @@ f 1
 :e
 f { a = 1 }
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.377: 	f { a = 1 }
+//│ ║  l.378: 	f { a = 1 }
 //│ ║         	^^^^^^^^^^^
 //│ ╟── operator application of type `int` does not have field 'a'
-//│ ║  l.347: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ║         	                                             ^^^^^
 //│ ╟── Note: constraint arises from field selection:
-//│ ║  l.347: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ╙──       	                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 //│ res: error
 //│    = { a: 1 }
@@ -400,13 +414,13 @@ f ainf
 //│ Runtime error:
 //│   ReferenceError: ainf is not defined
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.398: 	f ainf
+//│ ║  l.399: 	f ainf
 //│ ║         	^^^^^^
 //│ ╟── operator application of type `int` does not have field 'a'
-//│ ║  l.347: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ║         	                                             ^^^^^
 //│ ╟── Note: constraint arises from field selection:
-//│ ║  l.347: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ╙──       	                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 //│ res: error
 //│ Runtime error:
@@ -419,13 +433,13 @@ f infina
 //│ Runtime error:
 //│   ReferenceError: infina is not defined
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.417: 	f infina
+//│ ║  l.418: 	f infina
 //│ ║         	^^^^^^^^
 //│ ╟── operator application of type `int` does not have field 'a'
-//│ ║  l.347: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ║         	                                             ^^^^^
 //│ ╟── Note: constraint arises from field selection:
-//│ ║  l.347: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
+//│ ║  l.348: 	rec def f x = if x > 0 then (f (x with { a = x - 1 })).a else x
 //│ ╙──       	                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 //│ res: error | int | (('b | (0 & {a: 'c} as 'c) | 'a)\a & {a: int} as 'a)
 //│ Runtime error:
@@ -439,16 +453,16 @@ def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as
 :e
 f_manual 1
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.440: 	f_manual 1
+//│ ║  l.441: 	f_manual 1
 //│ ║         	^^^^^^^^^^
 //│ ╟── integer literal of type `1` does not have field 'a'
-//│ ║  l.440: 	f_manual 1
+//│ ║  l.441: 	f_manual 1
 //│ ║         	         ^
 //│ ╟── Note: constraint arises from record type:
-//│ ║  l.435: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
+//│ ║  l.436: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
 //│ ║         	                ^^^^^^^^^^^^^^^^^
 //│ ╟── from intersection type:
-//│ ║  l.435: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
+//│ ║  l.436: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
 //│ ╙──       	              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 //│ res: 1 | error | (('b | 1 | 'a)\a & {a: int} as 'a)
 //│    = <no result>
@@ -457,16 +471,16 @@ f_manual 1
 :e
 f_manual { a = 1 }
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.458: 	f_manual { a = 1 }
+//│ ║  l.459: 	f_manual { a = 1 }
 //│ ║         	^^^^^^^^^^^^^^^^^^
 //│ ╟── integer literal of type `1` does not have field 'a'
-//│ ║  l.458: 	f_manual { a = 1 }
+//│ ║  l.459: 	f_manual { a = 1 }
 //│ ║         	               ^
 //│ ╟── Note: constraint arises from record type:
-//│ ║  l.435: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
+//│ ║  l.436: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
 //│ ║         	                ^^^^^^^^^^^^^^^^^
 //│ ╟── from intersection type:
-//│ ║  l.435: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
+//│ ║  l.436: 	def f_manual: (({a: 'b & 'a & 'c} as 'a) & 'd) -> ('c | ('d | 'e\a & {a: int} as 'e))
 //│ ╙──       	                    ^^^^^^^^^^^^
 //│ res: 1 | error | ({a: 1} | 'a\a & {a: int} as 'a)
 //│    = <no result>
@@ -493,16 +507,16 @@ def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & ((
 :e
 f_manual_ns ainf
 //│ ╔══[ERROR] Type mismatch in application:
-//│ ║  l.494: 	f_manual_ns ainf
+//│ ║  l.495: 	f_manual_ns ainf
 //│ ║         	^^^^^^^^^^^^^^^^
 //│ ╟── type `int` does not have field 'a'
-//│ ║  l.489: 	def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & (({a: 'd & 'c} as 'c) | ~{a: 'e | int})\a & (({a: 'f} as 'c) as 'f) & (int | ~{a: 'e | int} | ~{})\a & (int | ~{a: 'e | int})\a & int & int) -> ('g | 'd | ('b | 'h\a & {a: 'e | int} as 'h))
+//│ ║  l.490: 	def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & (({a: 'd & 'c} as 'c) | ~{a: 'e | int})\a & (({a: 'f} as 'c) as 'f) & (int | ~{a: 'e | int} | ~{})\a & (int | ~{a: 'e | int})\a & int & int) -> ('g | 'd | ('b | 'h\a & {a: 'e | int} as 'h))
 //│ ║         	                                                              ^^^
 //│ ╟── Note: constraint arises from record type:
-//│ ║  l.489: 	def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & (({a: 'd & 'c} as 'c) | ~{a: 'e | int})\a & (({a: 'f} as 'c) as 'f) & (int | ~{a: 'e | int} | ~{})\a & (int | ~{a: 'e | int})\a & int & int) -> ('g | 'd | ('b | 'h\a & {a: 'e | int} as 'h))
+//│ ║  l.490: 	def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & (({a: 'd & 'c} as 'c) | ~{a: 'e | int})\a & (({a: 'f} as 'c) as 'f) & (int | ~{a: 'e | int} | ~{})\a & (int | ~{a: 'e | int})\a & int & int) -> ('g | 'd | ('b | 'h\a & {a: 'e | int} as 'h))
 //│ ║         	                              ^^^^^^^^^^^^
 //│ ╟── from intersection type:
-//│ ║  l.489: 	def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & (({a: 'd & 'c} as 'c) | ~{a: 'e | int})\a & (({a: 'f} as 'c) as 'f) & (int | ~{a: 'e | int} | ~{})\a & (int | ~{a: 'e | int})\a & int & int) -> ('g | 'd | ('b | 'h\a & {a: 'e | int} as 'h))
+//│ ║  l.490: 	def f_manual_ns: 'a | ('b & (({a: 'd & 'c} as 'c) | ~{a: 'e | int} | ~{})\a & (({a: 'd & 'c} as 'c) | ~{a: 'e | int})\a & (({a: 'f} as 'c) as 'f) & (int | ~{a: 'e | int} | ~{})\a & (int | ~{a: 'e | int})\a & int & int) -> ('g | 'd | ('b | 'h\a & {a: 'e | int} as 'h))
 //│ ╙──       	                                  ^^^^^^^
 //│ res: error
 //│    = <no result>
@@ -536,13 +550,13 @@ r + 1
 :e
 r.a
 //│ ╔══[ERROR] Type mismatch in field selection:
-//│ ║  l.537: 	r.a
+//│ ║  l.538: 	r.a
 //│ ║         	^^^
 //│ ╟── integer literal of type `1` does not have field 'a'
-//│ ║  l.526: 	r = f 1
+//│ ║  l.527: 	r = f 1
 //│ ║         	      ^
 //│ ╟── but it flows into reference with expected type `{a: ?a}`
-//│ ║  l.537: 	r.a
+//│ ║  l.538: 	r.a
 //│ ╙──       	^
 //│ res: error | int
 //│ Runtime error:

